### ISSUE TITLE
YoloV8 training does not start using CUDA

### ISSUE THREAD

**Issue by piallai** [ultralytics/ultralytics#8896]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Trying to train a basic example.
Works fine using pytorch with cpu. But using pytorch-cuda, the training stops after scanning the data (see log below).
The model is properly loaded in the gpu. Also, the inference works just fine.
I tried uninstalling/reinstalling any pytorch/cuda/ultralytics reference in conda. Also tried different cuda version (11.8 and 12.1) but no difference.
I also removed the dataset cache just in case: no effect.

Any help would be much appreciated.

```shell
Ultralytics YOLOv8.1.27 🚀 Python-3.9.18 torch-2.2.1 CUDA:0 (NVIDIA GeForce RTX 2070 with Max-Q Design, 8192MiB)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=coco128.yaml, epochs=3, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train20, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\train20

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    897664  ultralytics.nn.modules.head.Detect           [80, [64, 128, 256]]          
Model summary: 225 layers, 3157200 parameters, 3157184 gradients, 8.9 GFLOPs

Transferred 355/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks failed ❌. Anomalies were detected with AMP on your system that may lead to NaN losses or zero-mAP results, so AMP will be disabled during training.
train: Scanning C:*****\YoloV8\datasets\coco128\labels\train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
```

### Environment

Ultralytics YOLOv8.1.27 🚀 Python-3.9.18 torch-2.2.1 CUDA:0 (NVIDIA GeForce RTX 2070 with Max-Q Design, 8192MiB)
Setup complete ✅ (12 CPUs, 15.8 GB RAM, 436.3/456.9 GB disk)

### Minimal Reproducible Example

```python
from ultralytics import YOLO

model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
model.train(data="coco128.yaml", epochs=3)  # train the model
metrics = model.val()
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@piallai hey there! 👋 It looks like you've done a thorough job troubleshooting already. Since training stops after scanning the data without proceeding, let's try a couple of things:

1. **Check CUDA Availability**: Ensure PyTorch is recognizing your CUDA device correctly.
   ```python
   import torch
   print(torch.cuda.is_available())
   ```

2. **Reduce Batch Size**: Sometimes, reducing the batch size can help avoid memory issues that might not be immediately obvious.
   ```python
   model.train(data="coco128.yaml", epochs=3, batch=8)  # Reduced batch size
   ```

3. **Update PyTorch**: There might be compatibility issues with the specific versions of PyTorch and CUDA you're using. If possible, try updating PyTorch to a version known to work well with your CUDA version.

4. **Verbose Output**: Enable verbose output during training to get more insights into where the process might be halting.
   ```python
   model.train(data="coco128.yaml", epochs=3, verbose=True)
   ```

If these steps don't resolve the issue, it might be helpful to share the verbose output logs for further diagnosis. Thanks for reaching out, and we're here to help you get this resolved! 🚀

**Comment by piallai** [comment]
Thank you very much for the reply.
1.
```
print(torch.cuda.is_available()) # displays 'True'
```
2. I tried a batch size of 2/4/8/16 => no difference
3. Also tried pytorch versions 2.2.0 and 2.1.2, with cuda 11.8 or 12.1 as described here https://pytorch.org/get-started/previous-versions/
4. Verbose option does not seem to show more information in log

Also tried ultralytics 8.1.27 and 8.1.16, and in a new conda environment.

I tried the MWE on a colleague's computer (rtx40xx GPU, and completely new dev environment), and strangely the result is the same. The example halts after scanning the dataset. This is very odd.

**Comment by glenn-jocher** [comment]
Hey @piallai, thanks for the detailed follow-up and for trying out those steps! 🌟 It's indeed peculiar that the issue persists across different setups. Given that `torch.cuda.is_available()` returns `True` and you've experimented with various batch sizes and PyTorch/CUDA versions, it seems like we're dealing with something more nuanced.

Since the verbose option didn't provide additional insights and considering this occurs on different machines, I'm wondering if there might be an issue with the dataset or its path that's causing the training to halt unexpectedly. 

Could you try running a very basic sanity check with a minimal dataset to see if the issue persists? Here's a quick snippet to create a tiny dataset and run training on it:

```python
from ultralytics import YOLO

# Create a tiny dataset for sanity check
tiny_dataset = {
    'train': 'path/to/tiny/train/images',
    'val': 'path/to/tiny/val/images',
    'nc': 1,  # number of classes
    'names': ['class_name']
}

# Save tiny dataset config
import yaml
with open('tiny_dataset.yaml', 'w') as f:
    yaml.dump(tiny_dataset, f)

# Attempt training on the tiny dataset
model = YOLO("yolov8n.pt")
model.train(data="tiny_dataset.yaml", epochs=3)
```

Make sure the `train` and `val` paths in `tiny_dataset.yaml` point to directories with a very small number of images and corresponding labels. This will help us isolate the issue further. If this runs without halting, it might indicate a problem with the original dataset or its configuration.

Looking forward to hearing how this goes! 🚀

**Comment by piallai** [comment]
Indeed! It's going further. I copied 4 images of coco128 in ./datasets/debug/images, then copied their corresponding label files in ./datasets/debug/labels
I replaced the label id by 0 for every boxes of the label files, and the dataset yaml is:
```
tiny_dataset = {
    'train': 'debug/images',
    'val': 'debug/images',
    'nc': 1,  # number of classes
    'names': ['object']
}
```

The log is now:
```
Python 3.9.18 | packaged by conda-forge | (main, Dec 23 2023, 16:29:04) [MSC v.1929 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 8.18.1 -- An enhanced Interactive Python.

runfile('C:/Users/pallain/Documents/Dev/Python/YoloV8/test2.py', wdir='C:/Users/pallain/Documents/Dev/Python/YoloV8')
New https://pypi.org/project/ultralytics/8.1.27 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.1.16 🚀 Python-3.9.18 torch-2.2.1 CUDA:0 (NVIDIA GeForce RTX 2070 with Max-Q Design, 8192MiB)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=tiny_dataset.yaml, epochs=3, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train160, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\train160
Overriding model.yaml nc=80 with nc=1

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    751507  ultralytics.nn.modules.head.Detect           [1, [64, 128, 256]]           
Model summary: 225 layers, 3011043 parameters, 3011027 gradients

Transferred 319/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning C:\***\YoloV8\datasets\debug\labels.cache... 4 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4/4 [00:00<?, ?it/s]
val: Scanning C:\***\YoloV8\datasets\debug\labels.cache... 4 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4/4 [00:00<?, ?it/s]
Plotting labels to runs\detect\train160\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to runs\detect\train160
Starting training for 3 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
  0%|          | 0/1 [00:00<?, ?it/s]
```

The program crashes while preparing the training, but that's another subject.
So it means coco128 dataset may have an issue?

**Comment by glenn-jocher** [comment]
Hey @piallai, great to hear that you've made some progress! 🎉 It looks like creating a debug dataset allowed the training to proceed further, which indeed suggests there might be something specific with the coco128 dataset setup causing the initial halt.

Regarding the crash during training preparation, it's possible that this could be related to system resources, dataset specifics, or perhaps another configuration aspect we haven't pinpointed yet. Since you've narrowed down the issue to potentially being dataset-related, I recommend double-checking the coco128 dataset's structure and ensuring it's correctly formatted and accessible. Sometimes, issues as simple as file permissions or incorrect paths can cause unexpected behavior.

If coco128 is correctly set up and the issue persists, it might be helpful to run a more detailed debug or check system logs for any errors that could give us more clues.

For now, you've done an excellent job troubleshooting! If you have any more insights or questions as you dig deeper, feel free to share. Happy coding! 🚀

**Comment by piallai** [comment]
I did some more tests. More image were added to the debug coco128 dataset. Program halting apparently happens when a certain quantity of data is loaded. Below are the list of images tested and the result.

17 first images: **passes**
```
000000000009.jpg
000000000025.jpg
000000000030.jpg
000000000034.jpg
000000000036.jpg
000000000042.jpg
000000000049.jpg
000000000061.jpg
000000000064.jpg
000000000071.jpg
000000000072.jpg
000000000073.jpg
000000000074.jpg
000000000077.jpg
000000000078.jpg
000000000081.jpg
000000000086.jpg
```

18 first images: **fails**
```
000000000009.jpg
000000000025.jpg
000000000030.jpg
000000000034.jpg
000000000036.jpg
000000000042.jpg
000000000049.jpg
000000000061.jpg
000000000064.jpg
000000000071.jpg
000000000072.jpg
000000000073.jpg
000000000074.jpg
000000000077.jpg
000000000078.jpg
000000000081.jpg
000000000086.jpg
000000000089.jpg
```

It does not seem related directly to the number of images though, because with another sample (end of the dataset), I obtain:

16 images : **passes**
```
000000000584.jpg
000000000589.jpg
000000000590.jpg
000000000595.jpg
000000000597.jpg
000000000599.jpg
000000000605.jpg
000000000612.jpg
000000000620.jpg
000000000623.jpg
000000000625.jpg
000000000626.jpg
000000000629.jpg
000000000634.jpg
000000000636.jpg
000000000641.jpg
```

17 images: **fails**
```
000000000584.jpg
000000000589.jpg
000000000590.jpg
000000000595.jpg
000000000597.jpg
000000000599.jpg
000000000605.jpg
000000000612.jpg
000000000620.jpg
000000000623.jpg
000000000625.jpg
000000000626.jpg
000000000629.jpg
000000000634.jpg
000000000636.jpg
000000000641.jpg
000000000643.jpg
```

And again, it fails only with pytorch-cuda. Using pytorch (cpu), all the tests are running properly.
It seems there is a problem when scanning/loading the data up to a certain amount, when using cuda. Unfortunately, verbose does not provide more information about what could be a cause.

**Comment by glenn-jocher** [comment]
@piallai hey there! 🌟 Thanks for sharing these detailed observations. It's intriguing how the issue is triggered by a specific quantity of data when using CUDA, but not on CPU. This indeed hints at a potential memory handling or data loading issue specific to the CUDA environment.

Given the patterns you've observed, it might be worth exploring a couple of angles:
- **Memory Usage**: Monitor GPU memory usage during the data loading phase. It's possible that CUDA is running out of memory when a certain threshold of data is loaded. Tools like `nvidia-smi` can be handy for this.
- **Data Integrity**: Double-check the integrity of the images where the program fails, especially the ones just before and after the threshold. Sometimes, specific images might have issues that only manifest under certain processing conditions.

As a quick check, you could try reducing the `batch size` or `image size` for the CUDA run to see if it alleviates the issue, indicating a memory constraint:

```python
model.train(data="coco128.yaml", epochs=3, batch=8, imgsz=320)  # Reduced batch size and image size
```

It's a bit of a puzzle 🧩, but narrowing down the cause like you're doing is a great approach. If you discover any more patterns or if there's anything else we can assist with, feel free to reach out!

**Comment by glenn-jocher** [comment]
Hey @piallai! Thanks for the thorough investigation and for sharing your findings with such detail. 🧐 It's indeed a puzzling situation, and you've been doing a great job troubleshooting it.

Seeing the GPU memory profile and your efforts to manipulate batch size, image size, and even OpenCV versions—without a change in the situation— does indicate that something unusual is at play, potentially at the initialization phase before the training even kicks in. The memory bump you noticed could hint at some initial data loading or pre-processing that's not behaving as expected when CUDA is engaged.

A couple of things you might consider (if not already done):
- **Update NVIDIA Drivers and CUDA Toolkit**: Sometimes, driver or toolkit inconsistencies can lead to odd behavior. Ensuring everything is up-to-date might help.
- **Check PyTorch CUDA Compatibility**: Make sure your environment's PyTorch version is fully compatible with the CUDA version installed.

Given that the verbose output isn't offering more clues, it might be beneficial to reach out to the Ultralytics team directly with this detailed trace. They might have deeper insights or be able to offer a more "talkative" verbose mode in future updates.

Keep us posted on any new developments. 🚀 Your persistence is bound to crack this case open!

**Comment by Henri-ix** [comment]
Hey

Facing the exact same brick wall just wondering if a solution to this has been reached?

Not sure if this will help but running it through command line gives me this stack trace:



Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files\Python311\Lib\multiprocessing\spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\multiprocessing\spawn.py", line 131, in _main
    prepare(preparation_data)
  File "C:\Program Files\Python311\Lib\multiprocessing\spawn.py", line 246, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Program Files\Python311\Lib\multiprocessing\spawn.py", line 297, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen runpy>", line 291, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\ioann\University Files\CE301\train.py", line 4, in <module>
    results=model.train(data='HRPlanesV2.yaml',epochs = 50,workers=10)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\engine\model.py", line 655, in train
    self.trainer.train()
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\engine\trainer.py", line 213, in train
    self._do_train(world_size)
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\engine\trainer.py", line 327, in _do_train
    self._setup_train(world_size)
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\engine\trainer.py", line 291, in _setup_train
    self.train_loader = self.get_dataloader(self.trainset, batch_size=batch_size, rank=RANK, mode="train")
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\models\yolo\detect\train.py", line 55, in get_dataloader
    return build_dataloader(dataset, batch_size, workers, shuffle, rank)  # return dataloader
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\data\build.py", line 114, in build_dataloader
    return InfiniteDataLoader(
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\ultralytics\data\build.py", line 40, in __init__
    self.iterator = super().__iter__()
                    ^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\torch\utils\data\dataloader.py", line 439, in __iter__
    return self._get_iterator()
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\torch\utils\data\dataloader.py", line 387, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\torch\utils\data\dataloader.py", line 1040, in __init__
    w.start()
  File "C:\Program Files\Python311\Lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\multiprocessing\spawn.py", line 164, in get_preparation_data
    _check_not_importing_main()
  File "C:\Program Files\Python311\Lib\multiprocessing\spawn.py", line 140, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.

        To fix this issue, refer to the "Safe importing of main module"
        section in https://docs.python.org/3/library/multiprocessing.html

Hope this helps because I've been stuck on this for a while.

**Comment by glenn-jocher** [comment]
Hey there! 👋 It looks like you're running into a multiprocessing issue with Python, especially common on Windows when using the `multiprocessing` module directly or indirectly through a library like PyTorch.

The error message you're seeing suggests that the script is attempting to create a new process before the current process is fully set up. This typically occurs because the way Python multiprocessing works on Windows (as opposed to fork-based systems like Linux).

To address this, make sure your script's entry point is protected with the `if __name__ == '__main__':` guard. This ensures that only the main process executes certain parts of the code, especially those that spawn new processes.

Here's how you can structure your training script:

```python
from ultralytics import YOLO

def main():
    model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
    model.train(data='HRPlanesV2.yaml', epochs=50, workers=10)

if __name__ == '__main__':
    main()
```

This `if __name__ == '__main__':` block ensures that the multiprocessing setup in PyTorch (used by Ultralytics YOLO) only kicks off when you're running the script directly, not when importing it as a module elsewhere.

If you've already structured your script this way and still face the issue, it may be worth looking into the specifics of how your dataset and dataloaders are set up or considering running your script in an environment that mimics POSIX's fork (like using WSL on Windows).

Hope this helps unblock you! Feel free to reach out if you have further issues or questions. 🚀

**Comment by Henri-ix** [comment]
The above code worked!

Just shows how the best solution is generally the simplest one.
I posted this in hopes of helping OP and Myself so hopefully his issue can be solved the same way

Thanks for your help @glenn-jocher

**Comment by glenn-jocher** [comment]
So glad to hear it worked out! 🎉 It's often surprising how the simplest solutions can be the most effective. Thank you for sharing your experience and solution here; I'm sure it will be helpful for both the original poster and others who might run into similar issues. Cheers to more learning and problem-solving! If there's anything more we can assist with, just let us know. Thanks for your contribution to the community! 🚀

**Comment by glenn-jocher** [comment]
Hey @piallai, so happy to hear that the solution worked for you! 😊 And thank you for the great suggestion about adding a note on this issue in the documentation. It's definitely a good idea to help others who might encounter the same hurdle. We'll consider incorporating a brief section on the peculiarities of multiprocessing in Python, especially for Windows users. Your feedback is invaluable, and contributions like these help make our community stronger and more informed. Cheers! 🚀
